### PR TITLE
[Application] App control support

### PR DESCRIPTION
**Pull request by tiwanek**
Appcontrol part of application plugin were not implemented.
Introducing following function of application API:
 - Application.getRequestedAppControl()
 - RequestedApplicationControl.replyResult()
 - RequestedApplicationControl.replyFailure()
 - ApplicationManager.launchAppControl()
 - ApplicationManager.findAppControl()

BUG=XWALK-991

Requires:
 - https://review.tizen.org/gerrit/#/c/35215/
 - https://review.tizen.org/gerrit/#/c/35689/
 - https://review.tizen.org/gerrit/#/c/35700/

### CODE REVIEW

File: src/application/application_api.js [hunk 1]
```diff
@@ -1,4 +1,5 @@
 // Copyright (c) 2014 Intel Corporation. All rights reserved.
+// Copyright (c) 2014 Samsung Electronics Co., Ltd All Rights Reserved
```
Comment (halton) [inline src/application/application_api.js]: nit, 2015

File: src/application/application_api.js [hunk 2]
```diff
@@ -94,23 +105,181 @@ function defineReadOnlyProperty(object, key, value) {
   });
 }
 
-// Define Application interface, the getRequestedAppControl method will not be
-// implemented ATM.
+function defineNonNullProperty(object, key, value) {
+  var hvalue = value;
```
Comment (halton) [inline src/application/application_api.js]: nit, hvalue is not a meaningful name

File: src/application/application_api.js [hunk 3]
```diff
@@ -94,23 +105,181 @@ function defineReadOnlyProperty(object, key, value) {
...
+}
+
+// ApplicationControlData interface
+tizen.ApplicationControlData = function(key, value) {
+  if (!(this instanceof tizen.ApplicationControlData)) {
+    throw new TypeError();
+  }
+
+  if (typeof key !== 'string' ||
+      !(value instanceof Array)) {
```
Comment (halton) [inline src/application/application_api.js]: nit, squash into one line

File: src/application/application_api.js [hunk 4]
```diff
@@ -94,23 +105,181 @@ function defineReadOnlyProperty(object, key, value) {
...
+
+  defineNonNullProperty(this, 'key', key);
+  defineNonNullProperty(this, 'value', value);
+};
+
+function toApplicationControlDataArray(data) {
+  var array_objects = [];
+  for (var i = 0; i < data.length; i++) {
+    if (typeof data[i]['key'] != 'string' ||
+        !(data[i]['value'] instanceof Array)) {
```
Comment (halton) [inline src/application/application_api.js]: nit, squash into one line, you can check `tools/check-coding-style`, we use 100 chars

File: src/application/application_api.js [hunk 5]
```diff
@@ -94,23 +105,181 @@ function defineReadOnlyProperty(object, key, value) {
...
+      throw new tizen.WebAPIException(tizen.WebAPIException.TYPE_UNKNOWN_ERR);
+    }
+    array_objects.push(
+        new tizen.ApplicationControlData(
+            data[i]['key'], data[i]['value']));
```
Comment (halton) [inline src/application/application_api.js]: sqaush

File: src/application/application_api.js [hunk 6]
```diff
@@ -140,18 +309,19 @@ exports.getAppInfo = function(appId) {
...
+  if (result.error !== null && result.error !== undefined)
     throw new tizen.WebAPIException(result.error);
   return new ApplicationInformation(result.data);
 };
 
 exports.getAppsInfo = function(onsuccess, onerror) {
   if ((typeof onsuccess !== 'function') ||
-      (arguments.length > 1 && (onerror === undefined || typeof onerror !== 'function')))
+      (onerror !== null && onerror !== undefined &&
+       typeof onerror !== 'function'))
```
Comment (halton) [inline src/application/application_api.js]: squash

File: src/application/application_api.js [hunk 7]
```diff
@@ -180,19 +350,21 @@ exports.getAppContext = function(contextId) {
 
 exports.getAppsContext = function(onsuccess, onerror) {
   if ((typeof onsuccess !== 'function') ||
-      (arguments.length > 1 && (onerror === undefined || typeof onerror !== 'function')))
+      (onerror !== undefined && onerror !== null &&
+       typeof onerror !== 'function'))
```
Comment (halton) [inline src/application/application_api.js]: nit, sqaush

File: src/application/application_control.cc [hunk 8]
```diff
@@ -0,0 +1,289 @@
+// Copyright (c) 2014 Samsung Electronics Co., Ltd All Rights Reserved
```
Comment (halton) [inline src/application/application_control.cc]: 2015

File: src/application/application_control.cc [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+// FIXME: this is temporary hack for API that is private
+// this function signature matches the one in capi
+extern "C" int service_create_event(bundle* data, struct service_s** service);
+
+bool AddAppControlDataToService(const ApplicationControlData& data,
+    service_h service) {
+  const std::string& key = data.key();
+  const std::vector<std::string>& values = data.values();
+  std::unique_ptr<const char*[]> array_of_values(
+        new const char*[values.size()]);
```
Comment (halton) [inline src/application/application_control.cc]: nit, wrong indent

File: src/application/application_control.h [hunk 10]
```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2014 Samsung Electronics Co., Ltd All Rights Reserved
```
Comment (halton) [inline src/application/application_control.h]: nit

File: src/application/application_requested_control.cc [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+  std::unique_ptr<ApplicationRequestedControl> result(
+      new ApplicationRequestedControl(caller_id,
+          ApplicationControl::ApplicationControlFromService(service)));
+
+  service_destroy(service);
+  bundle_free(b);
+
+  return result;
+}
+
```
Comment (halton) [inline src/application/application_requested_control.cc]: nit, remove this line at end of flie

**Comment by tiwanek** [comment]
@halton

File: src/application/application_control.cc [hunk 12]
```diff
@@ -0,0 +1,289 @@
+// Copyright (c) 2015 Samsung Electronics Co., Ltd All Rights Reserved
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include <gio/gio.h>
+
+#include <algorithm>
+#include <cstdlib>
+
+#include "application/application_control.h"
```
Comment (halton) [inline src/application/application_control.cc]: nit, put self include before all other headers.

File: src/application/application_control.cc [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+
+  for (unsigned i = 0; i < values.size(); ++i)
+    array_of_values[i] = values[i].c_str();
+
+  return service_add_extra_data_array(service, key.c_str(),
+      array_of_values.get(), values.size()) == SERVICE_ERROR_NONE;
+}
+
+bool AddAppControlDataArrayToService(
+    const std::vector<std::unique_ptr<ApplicationControlData>>& array,
```
Comment (halton) [inline src/application/application_control.cc]: `>>` -> `> >`

File: src/application/application_control.cc [hunk 14]
```diff
@@ -0,0 +1,289 @@
...
+  return category_;
+}
+
+const std::vector<std::unique_ptr<ApplicationControlData>>&
+    ApplicationControl::app_control_data_array() const {
+  return app_control_data_array_;
+}
+
+bool ApplicationControl::ReplyResult(
+    const std::vector<std::unique_ptr<ApplicationControlData>>& data,
```
Comment (halton) [inline src/application/application_control.cc]: ditto

File: src/application/application_control.cc [hunk 15]
```diff
@@ -0,0 +1,289 @@
...
+    const std::string& encoded_bundle) const {
+  service_h service;
```
Comment (halton) [inline src/application/application_control.cc]: declare this var as close as first use

File: src/application/application_control.cc [hunk 16]
```diff
@@ -0,0 +1,289 @@
...
+      SERVICE_RESULT_SUCCEEDED) == SERVICE_ERROR_NONE;
+  service_destroy(reply);
+  service_destroy(service);
+  bundle_free(b);
+
+  return success;
+}
+
+bool ApplicationControl::ReplyFailure(const std::string& encoded_bundle) const {
+  service_h service;
```
Comment (halton) [inline src/application/application_control.cc]: ditto

File: src/application/application_control.cc [hunk 17]
```diff
@@ -0,0 +1,289 @@
...
+    category = c_category;
+    free(c_category);
+  }
+
+  std::unique_ptr<ApplicationControl> app_control(
+      new ApplicationControl(operation, uri, mime, category));
+
+  if (service_foreach_extra_data(s,
+      [](service_h service, const char* key, void* user_data) {
+        ApplicationControl* ac = static_cast<ApplicationControl*>(user_data);
```
Comment (halton) [inline src/application/application_control.cc]: declare this var as close as first used.

File: src/application/application_control.cc [hunk 18]
```diff
@@ -0,0 +1,289 @@
...
+        return true;
+      }, app_control.get()) != SERVICE_ERROR_NONE)
+    return nullptr;
+
+  return app_control;
+}
+
+std::unique_ptr<ApplicationControl>
+ApplicationControl::ApplicationControlFromJSON(
+    const picojson::value& value) {
```
Comment (halton) [inline src/application/application_control.cc]: squash above two lines

File: src/application/application_control.h [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+#include <vector>
+
+#include "application/application_control_data.h"
+#include "common/picojson.h"
+#include "common/utils.h"
+
+bool AddAppControlDataToService(const ApplicationControlData& data,
+    service_h service);
+bool AddAppControlDataArrayToService(
+    const std::vector<std::unique_ptr<ApplicationControlData>>& array,
```
Comment (halton) [inline src/application/application_control.h]: ditto

File: src/application/application_control.h [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+      const std::string& operation,
+      const std::string& uri,
+      const std::string& mime,
+      const std::string& category);
+
+  const std::string& operation() const;
+  const std::string& mime() const;
+  const std::string& uri() const;
+  const std::string& category() const;
+  const std::vector<std::unique_ptr<ApplicationControlData>>&
```
Comment (halton) [inline src/application/application_control.h]: ditto

File: src/application/application_control.h [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+      app_control_data_array() const;
+
+  bool ReplyResult(
+      const std::vector<std::unique_ptr<ApplicationControlData>>& data,
```
Comment (halton) [inline src/application/application_control.h]: ditto

File: src/application/application_control.h [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+      ApplicationControlFromService(service_h b);
+  static std::unique_ptr<ApplicationControl>
+      ApplicationControlFromJSON(const picojson::value& value);
+
+ private:
+  std::string operation_;
+  std::string mime_;
+  std::string uri_;
+  std::string category_;
+  std::vector<std::unique_ptr<ApplicationControlData>> app_control_data_array_;
```
Comment (halton) [inline src/application/application_control.h]: ditto

File: src/application/application_control_data.cc [hunk 23]
```diff
@@ -0,0 +1,59 @@
+// Copyright (c) 2015 Samsung Electronics Co., Ltd All Rights Reserved
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "application/application_control_data.h"
+
+ApplicationControlData::ApplicationControlData(
+    const std::string& key,
+    std::unique_ptr<std::vector<std::string>> values)
```
Comment (halton) [inline src/application/application_control_data.cc]: ditto

File: src/application/application_control_data.cc [hunk 24]
```diff
@@ -0,0 +1,59 @@
...
+  for (const auto& i : *values_) {
+    array.push_back(picojson::value(i));
+  }
+  object["value"] = picojson::value(array);
+  return std::unique_ptr<picojson::value>(new picojson::value(object));
+}
+
+std::unique_ptr<ApplicationControlData>
+ApplicationControlData::ApplicationControlDataFromJSON(
+      const picojson::value& value) {
```
Comment (halton) [inline src/application/application_control_data.cc]: 4-space indent

File: src/application/application_control_data.h [hunk 25]
```diff
@@ -0,0 +1,36 @@
...
+#include <vector>
+
+#include "common/picojson.h"
+#include "common/utils.h"
+
+class ApplicationControlData {
+ public:
+  ApplicationControlData(
+      const std::string& key,
+      std::unique_ptr<std::vector<std::string>> values);
```
Comment (halton) [inline src/application/application_control_data.h]: ditto

File: src/application/application_instance.cc [hunk 26]
```diff
@@ -187,8 +268,57 @@ void ApplicationInstance::HandleLaunchApp(const picojson::value& msg) {
...
+  }
+
+  std::unique_ptr<picojson::value> result(
+      extension_->app_manager()->FindAppControl(*control));
+  ReturnMessageAsync(GetJSCallbackId(msg), *result);
+}
+
 void ApplicationInstance::ReturnMessageAsync(double callback_id,
-                                             picojson::value& value) {
+    picojson::value& value) {
```
Comment (halton) [inline src/application/application_instance.cc]: i'd personally like to keep the old style

File: src/application/application_requested_control.cc [hunk 27]
```diff
@@ -0,0 +1,59 @@
+// Copyright (c) 2015 Samsung Electronics Co., Ltd All Rights Reserved
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include <app_service.h>
+#include <bundle.h>
+
+#include "application/application_requested_control.h"
```
Comment (halton) [inline src/application/application_requested_control.cc]: ditto, the header order

**Comment by halton** [comment]
And I bet you need update the test case as well.

File: src/application/application_control.cc (hunk 13 above)
Comment (tiwanek) [inline src/application/application_control.cc]: Allowed according to: http://chromium-cpp.appspot.com/

File: src/application/application_control.cc (hunk 14 above)
Comment (tiwanek) [inline src/application/application_control.cc]: Allowed according to: http://chromium-cpp.appspot.com/

File: src/application/application_control.h (hunk 19 above)
Comment (tiwanek) [inline src/application/application_control.h]: Allowed according to: http://chromium-cpp.appspot.com/

File: src/application/application_control.h (hunk 20 above)
Comment (tiwanek) [inline src/application/application_control.h]: Allowed according to: http://chromium-cpp.appspot.com/

File: src/application/application_control.h (hunk 21 above)
Comment (tiwanek) [inline src/application/application_control.h]: Allowed according to: http://chromium-cpp.appspot.com/

File: src/application/application_control.h (hunk 22 above)
Comment (tiwanek) [inline src/application/application_control.h]: Allowed according to: http://chromium-cpp.appspot.com/

File: src/application/application_control_data.cc (hunk 23 above)
Comment (tiwanek) [inline src/application/application_control_data.cc]: Allowed according to: http://chromium-cpp.appspot.com/

File: src/application/application_control_data.h (hunk 25 above)
Comment (tiwanek) [inline src/application/application_control_data.h]: Allowed according to: http://chromium-cpp.appspot.com/

**Comment by tiwanek** [comment]
https://crosswalk-project.org/jira/browse/XWALK-1490?focusedCommentId=28601&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-28601

**Comment by halton** [comment]
@tiwanek there is example to application api under `examples/application.html`.

**Comment by tiwanek** [comment]
About examples. I'm not sure what I should do about it.

Each function, I have added, bases its behaviour on other apps being installed on tizen (it searches for them or launch them).

If you have some suggestions, please, let me know. I will try to fix it on next PR.

**Comment by halton** [comment]
Initially, API owner write test case (example/api.html) along with  the implementation. There is tizen-extensions-crosswalk-examples.rpm will be built out if you run gbs build. We should keep the implementation and example aligned all the time.

**Comment by tiwanek** [comment]
After short look on examples. I will add "handleFindAppControl" and "handleLaunchAppControl" in separate PR.

**Comment by halton** [comment]
Fine with separate PR for the test case. LGTM and this PR will merging soon.

**Pull request merged**